src/fheroes2/dialog/dialog_adventure.cpp

```diff
@@ -21,106 +21,124 @@
+                break;
+            }
+            if ( le.MouseClickLeft( buttonPuzzle.area() ) || Game::HotKeyPressEvent( Game::HotKeyEvent::WORLD_PUZZLE_MAP ) ) {
+                result = Dialog::PUZZLE;
+                break;
+            }
+            if ( le.MouseClickLeft( buttonInfo.area() ) || Game::HotKeyPressEvent( Game::HotKeyEvent::WORLD_SCENARIO_INFORMATION ) ) {
+                back.restore();
+                result = Dialog::INFO;
+                break;
+            }
+            if ( ( le.MouseClickLeft( buttonDig.area() ) || Game::HotKeyPressEvent( Game::HotKeyEvent::WORLD_DIG_ARTIFACT ) ) && buttonDig.isEnabled() ) {
+                back.restore();
+                result = Dialog::DIG;
+                break;
+            }
+            if ( le.MouseClickLeft( buttonCancel.area() ) || Game::HotKeyCloseWindow() ) {
+                result = Dialog::CANCEL;
+                break;
+            }
+
+            // right info
+            if ( le.MousePressRight( buttonWorld.area() ) )
+                fheroes2::showStandardTextMessage( _( "View World" ), _( "View the entire world." ), Dialog::ZERO );
+            if ( le.MousePressRight( buttonPuzzle.area() ) )
+                fheroes2::showStandardTextMessage( _( "Puzzle" ), _( "View the obelisk puzzle." ), Dialog::ZERO );
+            if ( le.MousePressRight( buttonInfo.area() ) )
+                fheroes2::showStandardTextMessage( _( "Scenario Information" ), _( "View information on the scenario you are currently playing." ), Dialog::ZERO );
+            if ( le.MousePressRight( buttonDig.area() ) )
+                fheroes2::showStandardTextMessage( _( "Digging" ), _( "Dig for the Ultimate Artifact." ), Dialog::ZERO );
+            if ( le.MousePressRight( buttonCancel.area() ) )
+                fheroes2::showStandardTextMessage( _( "Cancel" ), _( "Exit this menu without doing anything." ), Dialog::ZERO );
         }
 
-        // right info
-        if ( le.MousePressRight( buttonWorld.area() ) )
-            fheroes2::showStandardTextMessage( _( "View World" ), _( "View the entire world." ), Dialog::ZERO );
-        if ( le.MousePressRight( buttonPuzzle.area() ) )
-            fheroes2::showStandardTextMessage( _( "Puzzle" ), _( "View the obelisk puzzle." ), Dialog::ZERO );
-        if ( le.MousePressRight( buttonInfo.area() ) )
-            fheroes2::showStandardTextMessage( _( "Scenario Information" ), _( "View information on the scenario you are currently playing." ), Dialog::ZERO );
-        if ( le.MousePressRight( buttonDig.area() ) )
-            fheroes2::showStandardTextMessage( _( "Digging" ), _( "Dig for the Ultimate Artifact." ), Dialog::ZERO );
-        if ( le.MousePressRight( buttonCancel.area() ) )
-            fheroes2::showStandardTextMessage( _( "Cancel" ), _( "Exit this menu without doing anything." ), Dialog::ZERO );
-    }
+        // restore background
+        back.restore();
+        display.render( background.totalArea() );
```

Comment:
In the loop we can render only the `activeArea` (the inner dialog area where all dialog items can change). The `totalArea` should be rendered only when dialog is opened and closed to render also the dialod borders and its shadow.
```suggestion
        display.render( background.activeArea() );
```

This comment is also for other changed dialogs.

Reply:
@zenseii, you are absolutely right with these changes!